packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts

```diff
@@ -73,4 +88,111 @@ export class ApplicationExtension<
     }
+
+    /**
+     * Called on the server to serialize the extension data that will be sent to the client.
+     *
+     * @returns SerializedExtension - The serialized extension data.
+     * @throws Error if the routes have not been loaded.
+     */
+    public serialize(): SerializedExtension {
+        if (this._cachedRoutes === null) {
+            throw new Error('Routes have not been loaded. Call getRoutes() before serializing')
+        }
+        const serializedRoutes = this._cachedRoutes.map((route) => {
+            if (!route.component?.displayName) {
+                throw new Error(
+                    `Component for route with path "${
+                        route.path
+                    }" is missing a displayName in ${this.getName()} extension`
+                )
+            }
+            return {
+                path: route.path,
+                componentName: route.component?.displayName
+            }
+        })
+        return {
+            routes: serializedRoutes
+        }
+    }
+
+    /**
+     * Returns a map of component names to components that are used to deserialize the extension data.
+     *
+     * @protected
+     * @returns ComponentMap - The map of component names to components.
+     */
+    protected getComponentMap?(): ComponentMap
+
+    /**
+     * Called on the client to deserialize the extension data that was serialized on the server.
+     *
+     * @param serializedExtension - The serialized extension data.
+     * @returns DeserializedExtension - The deserialized extension data.
+     * @throws Error if getComponentMap() is not defined.
+     * @throws Error if the deserialized component cannot be found in the component map.
+     */
+    private deserialize(serializedExtension: SerializedExtension): DeserializedExtension {
+        if (!this.getComponentMap) {
+            throw new Error(
+                `${this.getName()}.getRoutes() is async but does not define getComponentMap()`
+            )
+        }
+
+        const componentMap = this.getComponentMap()
+        const routes = serializedExtension.routes.map(({path, componentName}) => {
+            const component = componentMap[componentName]
+
+            if (!component) {
+                throw new Error(
+                    `${componentName} component could not be deserialized for route with path: ${path}`
+                )
+            }
+
+            return {
+                path,
+                exact: true,
+                component
+            }
+        })
+        return {routes}
+    }
+
+    private handleAsyncRoutes() {
+        if (!isServerSide) {
+            // Deserialize the routes on the client to ensure the latest routes are loaded on the client
+            this._cachedRoutes = this.deserialize(window.__EXTENSIONS__[this.getName()]).routes
+        }
+
+        // Apply caching for the getRoutes method
+        this.applyCacheForMethod('getRoutes', '_cachedRoutes')
+    }
```

Comment:
here we have logic specific to when an extension implements an async `getRoutes`.

I've only enabled caching for the `getRoutes` method when `getRoutes` is async.
Would we want it to cache even if `getRoutes` is sync? In other words, would we ever want to serialize the extension when getRoutes is sync?

Reply:
I don't see any reason why would want to serialize the extension if the getRoutes isn't async.. that being said we don't know what else the extension might do in the future, but for now it's a safe decision to make.